### ISSUE TITLE
Idea of implementation for a Kotlin-like copy function

### ISSUE THREAD

**Issue by MrJorgeXavier** [WilhelmOks/ModifiedCopyMacro#5]
Thanks for doing this package, it's very useful!

I saw in the readme that you would like to know some ideas on how to implement the copy function like kotlin does, and I think this issue could act as a thread to discuss how could we approach this.

I have an approach that uses Optional in a way that we can know if the caller wants to put nil in the property or keep the same value as before.
However, this approach has a drawback: The required values (not optional) of the struct will still be declared as Optional in the copy method, so the caller could pass nil there, expecting it to nullify the property, but it would keep the same value as before instead (since the property is not optional).

```swift 

struct MyStruct {
    let strValue: String?
    let intValue: Int

    /// Expanded copy method:
    /// If the type is optional, we assign the default value as `.some(nil)` instead of `nil`, so we can assert the case where the caller wants to put `nil` in the property.
    /// If the type is not optional, we assign the default value as `nil`, so we can know if the user wants to put a new value or keep the old one in the property.
    func copy(
        strValue: String?? = .some(nil),
        intValue: Int? = nil
    ) -> MyStruct {
        return MyStruct(
            strValue: getValue(strValue, currentValue: self.strValue),
            intValue: intValue ?? self.intValue
        )
    }
}

/// This function needs to be in an appropriate place... But I'm not sure where.
fileprivate func getValue<T>(_ doubleOptional: T??, currentValue: T?) -> Optional<T> {
    switch doubleOptional {
    case .none:
        return nil
    case .some(let wrapped):
        switch wrapped {
        case .none: return currentValue
        case .some(let newValue): return newValue
        }
    }
}

func tests() {
        let original = MyStruct(strValue: "Hello", intValue: 20)

        let copy1 = original.copy(strValue: nil, intValue: 10) // outputs strValue nil and intValue 10

        let copy2 = original.copy(strValue: "World") // outputs strValue "World" and intValue 20

        let copy3 = original.copy(intValue: 15) // outputs strValue "Hello" and intValue 15

        let copy4 = original.copy(intValue: nil) // outputs strValue "Hello" and intValue 20. This case is one that should not be possible to call, but unfortunately it is.
}
```

Let me know your opinions on this matter, I would be glad to open a pull request with this solution if we deemed it good enough.

**Comment by WilhelmOks** [comment]
Thank you for your suggestion and sorry for responding so late.

If I understand you correctly, you are suggesting to make everything optional and using nil as a special value, which would mean "keep the old value". 
This would mean that already optional values would become double optional.

I had this thought too, but unfortunately that would not be a satisfactory way to do it.
The reason is that, as you also have realized, there would be no possible way to distinguish between the intent of setting something to nil or actually to keep the old value. And that would be surprising and confusing.

**Comment by WilhelmOks** [comment]
I read your comment a second time and realized that I missed some points that you made.
So, we could actually distinguish between setting to nil and keeping the old value.
That would be nil vs. .some(nil).
This sounds like it could work.

My only worry is that I don't know if it will be possible for the macro to check that a property is optional, so that we pass it the default value of .some(nil) instead of nil for the non optional properties.
But maybe it's also the other way around and we will need to pass nil for "keep the old value" and .some(nil) for "set to nil". That would be easier to implement but it would be surprising on the call side, because the user would need to pass .some(nil) explicitly for "set to nil". Not intuitive.

I would like if you could try this and make a PR.

**Comment by WilhelmOks** [comment]
As I'm thinking about this now, I got another idea, which might solve all of the issues...

(I will omit the intValue property from your example to make it shorter.)

What if we declare a type that is similar to Optional<Some>

```
// names are to be decided later :D
enum Foo<T> {
  case unchanged
  case changed(_ newValue: T)
}
```

Then the generated copy function could look like this:

```
    func copy(
        strValue: Foo<String> = .unchanged
    ) -> MyStruct {
        return MyStruct(
            strValue: switch strValue {
            case .unchanged: self.strValue
            case .changed(let newValue): newValue
            }
        )
    }
```

The call side would look like this

```
  let original = MyStruct(strValue: "Hello")
  let copy1 = original.copy(strValue: .changed("Peter"))
```

and then we need to figure out how the Optional type does its magic so that we dont need to pass values to it like this
```
.some(newValue)
```
And then we can replicate this for the Foo type and use the function like this:

```
  let copy1 = original.copy(strValue: "Peter")
```

**Comment by WilhelmOks** [comment]
I have looked for a solution to make this possible:

```
let myValue: Foo<String> = "value" 
//rather than this:
let myValue: Foo<String> = .changed("value") 
```

I couldn't find anything online (I'm bad at googling), but chatgpt has made a suggestion:
```
enum Foo<T> {
    case unchanged
    case changed(T)
    
    init(_ newValue: T) {
        self = .changed(newValue)
    }
}
```

Not sure if that would work. I would need to try it out.

**Comment by WilhelmOks** [comment]
Fixed with #7
